src/views/SystemIntake/ContractDetails/ContractFields.tsx

```diff
@@ -0,0 +1,300 @@
+import React, { useEffect } from 'react';
+import { Controller } from 'react-hook-form';
+import { useTranslation } from 'react-i18next';
+import { ErrorMessage } from '@hookform/error-message';
+import { Fieldset, TextInput } from '@trussworks/react-uswds';
+
+import { useEasiFormContext } from 'components/EasiForm';
+import {
+  DateInputDay,
+  DateInputMonth,
+  DateInputYear
+} from 'components/shared/DateInput';
+import FieldErrorMsg from 'components/shared/FieldErrorMsg';
+import FieldGroup from 'components/shared/FieldGroup';
+import HelpText from 'components/shared/HelpText';
+import Label from 'components/shared/Label';
+import { ContractDetailsForm } from 'types/systemIntake';
+
+type ContractFieldsProps = {
+  id: string;
+};
+
+const ContractFields = ({ id }: ContractFieldsProps) => {
+  const { t } = useTranslation('intake');
+
+  const {
+    control,
+    register,
+    watch,
+    trigger,
+    formState: { errors, isSubmitted }
+  } = useEasiFormContext<ContractDetailsForm>();
+
+  const startDate = watch('contract.startDate');
+  const endDate = watch('contract.startDate');
+
+  const hasStartDateError: boolean = !!errors?.contract?.startDate;
+  const hasEndDateError: boolean = !!errors?.contract?.endDate;
+
+  // Trigger `startDate` validation when field is updated
+  useEffect(() => {
+    if (isSubmitted && hasStartDateError) {
+      trigger('contract.startDate');
+    }
+  }, [hasStartDateError, isSubmitted, startDate, trigger]);
+
+  // Trigger `endDate` validation when field is updated
+  useEffect(() => {
+    if (isSubmitted && hasEndDateError) {
+      trigger('contract.endDate');
+    }
+  }, [hasEndDateError, isSubmitted, endDate, trigger]);
+
+  return (
+    <div id={id} className="margin-left-4 margin-top-1 margin-bottom-2">
+      <FieldGroup
+        className="margin-top-0"
+        scrollElement="contract.contractor"
+        error={!!errors.contract?.contractor}
+      >
+        <Label htmlFor="contractor">{t('contractDetails.contractors')}</Label>
+        <ErrorMessage
+          errors={errors}
+          name="contract.contractor"
+          as={FieldErrorMsg}
+        />
+        <TextInput
+          {...register('contract.contractor')}
+          ref={null}
+          id="contractor"
+          type="text"
+          maxLength={100}
+        />
```

Comment:
RHF supports both controlled and uncontrolled form field components. Not sure about this specific case, but in general the UI lib we have fits with controlled fields, but there are instances where state handling (around stuff like validation or conditional fields) gets difficult and "uncontrolled" happens to be easier to fit in.

Reply:
React Hook Forms was made to primarily use uncontrolled components.  The `<Controller />` component is mainly for use with external libraries that don't expose the input ref ([see docs](https://www.react-hook-form.com/get-started/#IntegratingwithUIlibraries)). The Trussworks field components that we use all have an `inputRef` prop and can be used either controlled or uncontrolled, so we can use the `register` function instead of needing to wrap everything in a `<Controller />`.

I'll usually use `<Controller />` for more complicated fields. For example the [contact dropdown fields](https://github.com/CMS-Enterprise/easi-app/blob/main/src/views/SystemIntake/ContactDetails/index.tsx#L463-L495) in the system intake form need custom `value` and `onChange` props to transform the data.

So basically it's fine to use either the `register` function or wrap it in a `<Controller />`, but I think the `register` function is simpler with normal inputs because it's the intended functionality for RHF.

One thing to note is that since RHF relies a lot on refs, I did have to write a custom `register` function that returns an `inputRef` to work with the Trussworks components. So if you want to use `register` with a Trussworks component, the easiest way to do it is to use [useEasiForm](https://github.com/CMS-Enterprise/easi-app/blob/877a6dd45e022a9d455b98b80974dc12955dd180/src/components/EasiForm/useEasiForm.ts#L22) instead of `useForm`.

Reply:
oh wow interesting! thank you both for the context! 😄